### PR TITLE
`Programming exercise`: Reorder options for programming exercise creation

### PR DESCRIPTION
**Pull request by simon-hng**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context

Resolves https://github.com/ls1intum/Artemis/issues/6326

### Description

Moves settings during programming exercise configuration.

As the list given in the PR was not comprehensive for all configuration items and not definitely agreed on, I did the following changes:

- [x] Refactor all step inputs into a single config object
- [x] Create new `Infrastructure` step
- [x] Move preview repositories from `General Information` -> `Infrastructure`
- [x] Move Checkboxes `Problem Statement` -> `Infrastructure`
  - Show Test Names to Students
  - Publish Build Plan
- [x] Move Checkboxed `Problem Statement` -> `Language`
  - sequential test runs
  - record testwise coverage
  - static code analysis
- [x] Add wizard mode headline and descriptions to normal mode
- [x] Split SCA into checkbox and grading settings, keep checkbox where is and  and move penalty into grading
- [x] Move `Should this exercise be included in the course score calculation` to the top including points and bonus points.
- [x] Rename `Difficulty` into `Mode`
- [x] Move `Allow Offline IDE`, `Allow Online Editor` into `Mode`
- [x] Rename `Mode` setting into `Participation`
- [x] Hide `Submission policy` editing alert when creating new exercise (Only visible in edit)
- [x] Move `Assessment Instructions` into grading

### Steps for Testing

Prerequisites:
- 1 Instructor

1. Go to exercise creation
2. Create various exercises with different settings, switch both the normal as well as guided modes
3. Check that the exercise is created successfully with the correct settings after clicking the "finish" button for exercise creation 

#### Performance Review
- [x] I confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Screenshots

localhost_9000_course-management_1_programming-exercises_new
localhost_9000_course-management_1_programming-exercises_new (1)

### CODE REVIEW

**Comment by undernagruzez** [comment]
Tested on ts5. 
Normal mode: Not sure what the behavior should be. Could not tick some checkboxes. 
image
Guided mode: could not press the Next button
image

**Comment by simon-hng** [comment]
> Tested on ts5. Normal mode: Not sure what the behavior should be. Could not tick some checkboxes. image Guided mode: could not press the Next button image

Thanks for the review! It seems some issues came up during merging, which I fixed now.

The checkboxes are disabled until you set a due date or set (automatic/manual grading)
Could you test again and see if everything is okay now?

**Review by undernagruzez — APPROVED** [review]
Retested on ts6, works as expected now

**Review by SolizerCodes — COMMENTED** [review]
grafik
When selecting "Sequential Test Runs", "Record Testwise Coverage" looks like it can be selected but can't.

Apart from that looks good

**Comment by simon-hng** [comment]
> grafik When selecting "Sequential Test Runs", "Record Testwise Coverage" looks like it can be selected but can't.
> 
> Apart from that looks good

This PR is only intended to move the components around. I will fix this in a follow-up

**Review by TheZoker — DISMISSED** [review]
Tested within testing session on TS4. LGTM 👍

**Review by pal03377 — APPROVED** [review]
Tested in testing session on TS4. Could not find any issues 👍

**Review by FelixTJDietrich — DISMISSED** [review]
Tested on TS4 during a testing session. Everything worked as expected, I tested it extensively with most configurations and functionality works. Approved!

**Review by TheZoker — APPROVED** [review]

**Comment by FelixTJDietrich** [comment]
Reapproved after merge conflict where one attribute was added 👍

**Review by FelixTJDietrich — APPROVED** [review]
Reapproved after merge conflict where one attribute was added 👍

**Review by pal03377 — APPROVED** [review]
Re-approve after merge conflict

**Review by undernagruzez — APPROVED** [review]
Reapproved

File: src/main/webapp/app/exercises/programming/manage/update/programming-exercise-creation-config.ts
Comment (tobias-lippert) [inline src/main/webapp/app/exercises/programming/manage/update/programming-exercise-creation-config.ts:53]: can you move the isImportFromFile up to importFromExistingExercise, so related items are closer together. This makes it easier to understand and search for specific options

File: src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts
Comment (tobias-lippert) [inline src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts:1004]: same as above, can you please group the import related parameters closer together

**Review by tobias-lippert — COMMENTED** [review]
Looks good in general, just two minor remarks.

**Review by tobias-lippert — APPROVED** [review]
approve, comments will be addressed in a follow-up.

**Review by JohannesStoehr — APPROVED** [review]

**Review by Mtze — APPROVED** [review]

**Review by b-fein — APPROVED** [review]
Looks good. Thanks for implementing the changes!

(Forgot to actually submit the review before.)

**Pull request merged**
